File: src/typescript/openshell-driver.ts
/**
 * OpenShellDriver: resolves to OpenShellGrpcDriver.
 */

import { execSync } from "child_process";
import {
  type ShellDriver,
  ShellDriverFactory,
  type ShellDriverOptions,
} from "./drivers.js";
import {
  OpenShellGrpcDriver,
  type OpenShellGrpcDriverOptions,
} from "./openshell-grpc-driver.js";

export class OpenShellDriver {
  /**
   * Return an OpenShell ShellDriver backed by gRPC (via SSH for sync exec).
   */
  static resolve(opts: OpenShellGrpcDriverOptions = {}): ShellDriver {
    // If test override provided, use it directly
    if (opts._execOverride) {
      return new OpenShellGrpcDriver(opts);
    }
    // Check if SSH is available (used for sync exec in Node.js)
    if (!OpenShellDriver._checkSsh()) {
      throw new Error(
        "ssh not found — OpenShell driver requires SSH for synchronous execution"
      );
    }
    return new OpenShellGrpcDriver(opts);
  }

  /** Check if ssh is on PATH. */
  private static _checkSsh(): boolean {
    try {
      execSync("which ssh", { stdio: "ignore" });
      return true;
    } catch {
      return false;
    }
  }
}

/**
 * Register the "openshell" driver with ShellDriverFactory.
 */
export function registerOpenShellDriver(): void {
  ShellDriverFactory.register("openshell", (opts?: ShellDriverOptions) =>
    OpenShellDriver.resolve(opts as OpenShellGrpcDriverOptions)
  );
}

File: src/python/openshell_driver.py
"""OpenShellDriver: resolves to OpenShellGrpcDriver when grpcio is available."""

from __future__ import annotations

from typing import Any

from src.python.drivers import ShellDriver, ShellDriverFactory


class OpenShellDriver:
    """Resolves the openshell driver (requires grpcio package)."""

    @staticmethod
    def resolve(**kwargs: Any) -> ShellDriver:
        """Return an OpenShellGrpcDriver if grpcio is installed."""
        # Allow test override to skip import check
        grpc_override = kwargs.get("grpc_override")
        if grpc_override is not None:
            from src.python.openshell_grpc_driver import OpenShellGrpcDriver
            return OpenShellGrpcDriver(**kwargs)
        try:
            import grpc  # noqa: F401
        except ImportError:
            raise RuntimeError(
                "grpcio not found — install with: pip install agent-harness[openshell]"
            )
        from src.python.openshell_grpc_driver import OpenShellGrpcDriver
        return OpenShellGrpcDriver(**kwargs)


def register_openshell_driver() -> None:
    """Register the 'openshell' driver with the ShellDriverFactory."""
    ShellDriverFactory.register("openshell", lambda **kw: OpenShellDriver.resolve(**kw))
